src/settings.cpp

```diff
@@ -1018,3 +1089,259 @@ void Settings::doCallbacks(const std::string &name) const
+
+	if ((category & SETTINGS_ENFORCED))
+		empty = !serializeSettings(m_server_enforced, os, flags) && empty;
+
+	return !empty;
+}
+
+
+// Serialization of a group's contents: aggregate all settings into category 'customized'
+// Returns true if anything was serialized
+bool Settings::serializeGroup(std::ostream &os, Flags flags) const
+{
+	MutexAutoLock lock(m_mutex);
+	bool empty = true;
+
+	writeU8(os,SETTINGS_CUSTOMIZED);
+	empty = !serializeCategory(os, SETTINGS_CATEGORY_ALL, flags) && empty;
+	serializeSentinel(os, "CUSTOMIZED");
+
+	writeU8(os,SETTINGS_CATEGORY_NONE);
+	serializeSentinel(os, "GROUP");
+
+	return !empty;
+}
+
+
+// Regular serialization: serialize exactly what we've got.
+// Returns true if anything was serialized
+//
+// Serialization format (same as the format of a group):
+//	u8			category
+//	<string,u8,string>*	list of name, value-type, value
+//	<string,u8,string>	sentinel
+//	(repeat the 3 above as often as necessary)
+//	u8			category (SETTINGS_CATEGORY_NONE)
+//	<string,u8,string>	group sentinel
+// A value can itself be a group
+// ATM groups are assumed to consist of a single category only.
+bool Settings::serialize(std::ostream &os) const
+{
+	MutexAutoLock lock(m_mutex);
+	bool empty = true;
+
+	// Categories are terminated by a sentinel. Identified by name == "" and U8 == SERIALIZATION_VALUETYPE_SENTINEL
+	// Groups are also terminated by a group sentinel
+
+	writeU8(os,SETTINGS_DEFAULT);
+	empty = !serializeCategory(os, SETTINGS_DEFAULT, FLAGS_NONE) && empty;
+	serializeSentinel(os, "DEFAULT");
+
+	writeU8(os,SETTINGS_SUGGESTED);
+	empty = !serializeCategory(os, SETTINGS_SUGGESTED, FLAGS_NONE) && empty;
+	serializeSentinel(os, "SUGGESTED");
+
+	writeU8(os,SETTINGS_CUSTOMIZED);
+	empty = !serializeCategory(os, SETTINGS_CUSTOMIZED, FLAGS_NONE) && empty;
+	serializeSentinel(os, "CUSTOMIZED");
+
+	writeU8(os,SETTINGS_ENFORCED);
+	empty = !serializeCategory(os, SETTINGS_ENFORCED, FLAGS_NONE) && empty;
+	serializeSentinel(os, "ENFORCED");
+
+	writeU8(os,SETTINGS_CATEGORY_NONE);
+	serializeSentinel(os, "GROUP");
+
+	return !empty;
+}
+
+
+// For client; category depends on flags instead of on server category
+// Returns true if anything was serialized
+bool Settings::serializeForClient(std::ostream &os) const
+{
+	MutexAutoLock lock(m_mutex);
+	bool empty = true;
+
+	// Categories are terminated by a sentinel. Identified by name == "" and U8 == SERIALIZATION_VALUETYPE_SENTINEL
+
+	writeU8(os,SETTINGS_SUGGESTED);
+	empty = !serializeCategory(os, SETTINGS_CATEGORY_ALL, SUGGEST_TO_CLIENT) && empty;
+	serializeSentinel(os, "SUGGESTED");
+
+	writeU8(os,SETTINGS_ENFORCED);
+	empty = !serializeCategory(os, SETTINGS_CATEGORY_ALL, ENFORCE_ON_CLIENT) && empty;
+	serializeSentinel(os, "ENFORCED");
+
+	writeU8(os,SETTINGS_CATEGORY_NONE);
+	serializeSentinel(os, "GROUP");
+
+	return !empty;
+}
+
+
+// Must be entered with lock (m_mutex) held
+void Settings::deSerializeSettings(std::istream &is, SettingEntries *settings)
```

Comment:
you should not use this sentinel, a count of messages in group or global list is sufficient